src/Workspaces/Core/Portable/Storage/SQLite/v1/Interop/SafeSqliteStatementHandle.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using SQLitePCL;
+
+namespace Microsoft.CodeAnalysis.SQLite.v1.Interop
+{
+    internal sealed class SafeSqliteStatementHandle : SafeHandle
+    {
+        private readonly SafeHandleLease _lease;
+        private readonly sqlite3_stmt _wrapper;
+
+        public SafeSqliteStatementHandle(SafeSqliteHandle handle, sqlite3_stmt rawStatement)
+            : base(IntPtr.Zero, ownsHandle: true)
+        {
+            _lease = handle.Lease();
+            _wrapper = rawStatement;
+            SetHandle(rawStatement.ptr);
+        }
+
+        public override bool IsInvalid => handle == IntPtr.Zero;
+
+        public new sqlite3_stmt DangerousGetHandle()
+            => _wrapper;
+
+        protected override bool ReleaseHandle()
+        {
+            using var _ = _lease;
+
+            raw.sqlite3_finalize(_wrapper);
+            return true;
```

Comment:
given the duplication in hte pattern here, i wonder if it's worth having a base type that has the common logic, but which basically just forces derived type to implement the one call to actually release the udnerlying resource.

Reply:
`ReleaseHandle` _is_ the abstraction.

Reply:
sure.  but all impls need to do the lease work to ensure they dipose after.  I would ike that to be pulled up to common layer, and derived don't care about that, they only make the one call to actually release the resource.

Reply:
I prefer the current form. Each of the three handles in this pull request has distinct semantics. Any relation between them is coincidental, and I don't want to confuse this by trying to force them to share code.

Reply:
I ended up updating this to abstract the concept of a "child" handle.